### ISSUE TITLE
wrong call to --ppx flags (slash issue)

### ISSUE THREAD

**Issue by Et7f3** [ocaml/dune#5378]
## Expected Behavior

We should be able to execute test of [ppx_import](https://github.com/ocaml-ppx/ppx_import/tree/44c79bd97b7c8d901914baa7562174371ff5357d)

## Actual Behavior

We can't compile test

## Reproduction

- PR with a reproducing test: not yet

1. clone
2. dune runtest

## Specifications

- Version of `dune` (output of `dune --version`): https://github.com/ocaml/dune/commit/ef1ee9a0d175bb076d0debf917f7c67a3939af6a
- Version of `ocaml` (output of `ocamlc --version`): 4.13
- Operating system (distribution and version): windows

## Additional information 

I tried to debug:
This line seem the culprit https://github.com/ocaml/dune/blob/78fa5d22c3cc09345bfe3eb25fb57e0cd7e3c4bc/src/dune_rules/preprocessing.ml#L293

The command is: `ocamldep.opt.exe -modules -ppx ".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe --as-ppx" -impl src_test/ppx_deriving_sexp/test_ppx_deriving_sexp.ml`

Since -ppx go through [`Sys.command`](https://github.com/ocaml/ocaml/blob/8d394a670b81dcff39668b781b7b931cf6b2128c/driver/pparse.ml#L31). So we need to not normalize:
`ocamldep.opt.exe -modules -ppx ".ppx\57e0400f3b219ff4fc65d1fbd820e35d\ppx.exe --as-ppx" -impl src_test/ppx_deriving_sexp/test_ppx_deriving_sexp.ml` work as intended.

I tried to replace with: `Path.Build.L.relative build_dir (".ppx" :: key :: "ppx.exe" :: [])` but `Path` module seem to like normalized path `/`:
https://github.com/ocaml/dune/blob/ef1ee9a0d175bb076d0debf917f7c67a3939af6a/otherlibs/stdune/path.ml#L253

Have you any pointer to continue ?

- Link to gist with verbose output (run `dune` with the `--verbose` flag):

**Comment by ManasJayanth** [comment]
Alternate phrasing for the issue description (as experienced myself) - `staged_pps` fails to work inside Cygwin where Dune uses cmd.exe to load the ppx but uses Unix paths.

Tangential point: Can Dune look into $env:SHELL and use it instead while shelling out to commands?

**Comment by ManasJayanth** [comment]
I'm using the following patch (in the `Path.reach` function`) to unblock myself on Windows

```diff

  
--- ./src/stdune/path.ml
+++ ./src/stdune/path.ml
@@ -391,7 +391,7 @@
       | _ -> (
         match List.fold_left from ~init:t ~f:(fun acc _ -> ".." :: acc) with
         | [] -> "."
-        | l -> String.concat l ~sep:"/")
+        | l -> String.concat l ~sep:(if Sys.unix then "/" else "\\"))
     in
     loop (to_list t) (to_list from)
 
 ```
 
 Would really appreciate any pointers to the right direction. I'd be happy to PR this myself.

**Comment by nojb** [comment]
Thanks for the PR. Your analysis of the problem seems accurate. But am not sure about the best solution: I am afraid that changing  from forward slashes to backward slahses in `Path` may have unintended consequences (I recall that forward slashes are generally OK under Windows, here the issue has to do more with a `cmd.exe` peculiarity). Another possible approach could be to normalize to *backward* slashes under Windows only the argument of the `-ppx` flag (need to grep to find the place where this flag is passed).

> Tangential point: Can Dune look into $env:SHELL and use it instead while shelling out to commands? This can help us around this issue if guess the path separator turns out to be tricky.

This wouldn't help here, because it is the compiler that is shelling out the command in question, not `dune`. But in general this doesn't seem a good idea because it means the behaviour of `dune` (or the compiler) would be different depending on whether it is executed under Cygwin or not.

**Comment by ManasJayanth** [comment]
> Another possible approach could be to normalize to backward slashes under Windows only the argument of the -ppx flag (need to grep to find the place where this flag is passed).

That would be src/dune_rules/preprocessing.ml I think.

I tried this but noticed that Dune uses `Path.Build.t` there, and it canonicalised all paths to use forward slashes. I also recall a PR that marks paths with `\\` as not canonical and forces it through the normalisation again.

I agree that fix isn't as simple as changed `/` to `\\` everywhere :/

**Comment by ManasJayanth** [comment]
Is there a way to ask `Path.Build` to craft a path that contains `\\`?

**Comment by nojb** [comment]
> I tried this but noticed that Dune uses `Path.Build.t` there, and it canonicalised all paths to use forward slashes. I also recall a PR that marks paths with `\\` as not canonical and forces it through the normalisation again.

Yes, but at some point the `Path.Build.t` has to be converted back to a string, right? And we could do a purely textual replacement of forward slashes to backward slashes at that point (but it is true that is not very nice).

> Is there a way to ask `Path.Build` to craft a path that contains `\\`?

I don't recall at the moment but perhaps the slashes are preserved for *absolute* paths (see `Path.External`).

**Comment by nojb** [comment]
Another interesting point: I see that I can use forward slashes in `cmd.exe` if I surround the program name by double quotes, eg `C:\> "foo/bar.exe" arg1 arg2 ...`. Maybe it is just a matter of quoting the argument to `-ppx` ?

**Comment by ManasJayanth** [comment]
> Yes, but at some point the Path.Build.t has to be converted back to a string, right? And we could do a purely textual replacement of forward slashes to backward slashes at that point (but it is true that is not very nice).

My patch in https://github.com/ocaml/dune/issues/5378#issuecomment-1021021008 is exactly this. I realised after a lot of fighting with `Path.Build` that it's the only way to get backward slashes.

Also, very important to note: this happens only with `staged_pps` - I wonder why does `pps` work correctly, but not `staged_pps`?

**Comment by nojb** [comment]
> I wonder why does `pps` work correctly, but not `staged_pps`?

This is because `pps` is implemented more efficiently and does not go through the `-ppx` flag.

**Comment by nojb** [comment]
> then implement staged_pps like pps can be a viable solution ?

No, unfortunately, this is not possible (this is why `staged_pps` exists as as a different feature than `pps`).

> I looked for other workaround:

Did you try quoting the program name passed to `-ppx` between double quotes? That seems to work in a small experiment (see https://github.com/ocaml/dune/issues/5378#issuecomment-1021045121).

**Comment by Et7f3** [comment]
I don't managed to nest quoting:
```shell
> ocaml
        OCaml version 4.13.1

# #install_printer print_endline;;
# Filename.quote_command  "ocaml" ["-ppx"; "\".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe\" --as-ppx"];;
"ocaml ^"-ppx^" ^"\^".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe\^" --as-ppx^""
- : string =
# Filename.quote_command  "ocaml" ["-ppx"; "\".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe\" --as-ppx"] |> Sys.command;;
        OCaml version 4.13.1

# 1
  ;;
'.ppx' n’est pas reconnu en tant que commande interne
ou externe, un programme exécutable ou un fichier de commandes.
Line 1:
Error: Error while running external preprocessor
Command line: ".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe" --as-ppx "C:\Users\Et7f3\AppData\Local\Temp\camlppxb70a35" "C:\Users\Et7f3\AppData\Local\Temp\camlppx9410cd"

# ^Z

- : int = 0
# ^Z


> ocaml -ppx "\".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe\" --as-ppx"
        OCaml version 4.13.1

# 1;;
'.ppx' n’est pas reconnu en tant que commande interne
ou externe, un programme exécutable ou un fichier de commandes.
Line 1:
Error: Error while running external preprocessor
Command line: ".ppx/57e0400f3b219ff4fc65d1fbd820e35d/ppx.exe" --as-ppx "C:\Users\Et7f3\AppData\Local\Temp\camlppx6469c9" "C:\Users\Et7f3\AppData\Local\Temp\camlppxd59bee"

# ^Z
```

I think it is the time I need to summon @dra27

**Comment by rgrinberg** [comment]
@nojb do you think your spawn fix addressed this issue?

**Comment by nojb** [comment]
> @nojb do you think your spawn fix addressed this issue?

I don't know to be honest. @Et7f3 are you able to test your code with the tip of `main` and see if the problem persists? If not, I can try reproducing the problem and testing it myself, but it will take me a few days at least.

**Comment by Et7f3** [comment]
Will test it.

**Comment by Et7f3** [comment]
Issue reproduced with:

```
{
  "dependencies": {
    "@opam/dune": "3.1.1",
    "@opam/ounit": "2.2.6",
    "@opam/ppx_deriving": "5.2.1",
    "@opam/ppx_sexp_conv": "v0.15.0",
    "@opam/ppxlib": "0.25.0",
    "ocaml": "^4.14.0"
  },
  "resolutions": {
    "@opam/dune":
      "ocaml/dune:dune.opam#b6299ec743c71c762a25caa4e66f72e590267a42"
  }
}
```
just clone https://github.com/ocaml-ppx/ppx_import
copy this file as esy.json inside
and run `esy b dune runtest -p name`

**Comment by nojb** [comment]
cc @davesnx who recently had a problem involving `esy` and `-pp` that sounded quite similar (see #7235)

**Comment by nojb** [comment]
(I'm unnassigning myself because I am not sure what is the problem nor what could be done to move this issue forward, or even if it is still relevant.)
